### PR TITLE
LiBai ViT Profile [Not Merge]

### PR DESCRIPTION
**Pull request by rentainhe**
## ViT Profile分支
- 此分支主要对数据部分进行profile, 使用开启`OneDNN`编译的`Master`分支进行测试, 目前观察到的主要现象是:

| Data | Throughput |
|:---:|:---:|
| 开启Normalize | 1000左右  |
| 关闭Normalize | 1600-1800 |

## 运行方式 （在类脑上的使用）
- 8卡数据并行
```python
bash tools/train.sh configs/vit_imagenet.py 8
```
- 执行后的输出如下:
```
[03/07 09:29:56 lb.utils.events]:  iteration: 1/3002736  consumed samples: 128  total_loss: 1  data_time: 0.6959  lr: 5.00e-07  
[03/07 09:29:56 lb.utils.events]:  eta: 22:44:20  iteration: 2/3002736  consumed samples: 256  total_loss: 1  data_time: 0.3602  lr: 5.02e-07  
[03/07 09:29:56 lb.utils.events]:  eta: 10:24:51  iteration: 3/3002736  consumed samples: 384  total_loss: 1  time: 0.0125(10251.77)  data_time: 0.2442  lr: 5.05e-07  
[03/07 09:29:56 lb.utils.events]:  eta: 10:01:23  iteration: 4/3002736  consumed samples: 512  total_loss: 1  time: 0.0120(10651.70)  data_time: 0.1859  lr: 5.07e-07  
[03/07 09:29:57 lb.utils.events]:  eta: 10:24:51  iteration: 5/3002736  consumed samples: 640  total_loss: 1  time: 0.1585(807.56)  data_time: 0.2389  lr: 5.10e-07  
[03/07 09:29:57 lb.utils.events]:  eta: 14:39:56  iteration: 6/3002736  consumed samples: 768  total_loss: 1  time: 0.1245(1027.73)  data_time: 0.2028  lr: 5.12e-07  
[03/07 09:29:57 lb.utils.events]:  eta: 18:55:01  iteration: 7/3002736  consumed samples: 896  total_loss: 1  time: 0.1042(1228.00)  data_time: 0.1769  lr: 5.15e-07  
[03/07 09:29:57 lb.utils.events]:  eta: 16:51:40  iteration: 8/3002736  consumed samples: 1024  total_loss: 1  time: 0.0898(1425.07)  data_time: 0.1570  lr: 5.17e-07  
[03/07 09:29:57 lb.utils.events]:  eta: 18:55:01  iteration: 9/3002736  consumed samples: 1152  total_loss: 1  time: 0.1384(924.83)  data_time: 0.1872  lr: 5.20e-07  
[03/07 09:29:57 lb.utils.events]:  eta: 16:51:40  iteration: 10/3002736  consumed samples: 1280  total_loss: 1  time: 0.1225(1044.74)  data_time: 0.1696  lr: 5.22e-07  
[03/07 09:29:57 lb.utils.events]:  eta: 14:48:18  iteration: 11/3002736  consumed samples: 1408  total_loss: 1  time: 0.1106(1157.34)  data_time: 0.1555  lr: 5.25e-07  
[03/07 09:29:57 lb.utils.events]:  eta: 13:45:26  iteration: 12/3002736  consumed samples: 1536  total_loss: 1  time: 0.1005(1273.02)  data_time: 0.1434  lr: 5.27e-07
```
其中time后的括号内表示throughput的值，例如`time: 0.1384(924.83)`表示throughput为`924.83`, 可以多运行一阵子观测，然后取均值
- 如果需要修改transform, 比如关闭Normalize，需要去`/configs/common/data/imagenet.py`下对相应代码进行注释
```python
train_aug = LazyCall(transforms.Compose)(
    transforms=[
        LazyCall(transforms.RandomResizedCrop)(
            size=224,
            scale=(0.08, 1.0),
            ratio=(3.0 / 4.0, 4.0 / 3.0),
            interpolation=str_to_interp_mode("bicubic"),
        ),
        LazyCall(transforms.RandomHorizontalFlip)(p=0.5),
        LazyCall(rand_augment_transform)(
            config_str="rand-m9-mstd0.5-inc1",
            hparams=dict(
                translate_const=int(224 * 0.45),
                img_mean=tuple([min(255, round(255 * x)) for x in IMAGENET_DEFAULT_MEAN]),
                interpolation=str_to_interp_mode("bicubic"),
            ),
        ),
        LazyCall(transforms.ToTensor)(),
        # LazyCall(transforms.Normalize)(
        #     mean=IMAGENET_DEFAULT_MEAN,
        #     std=IMAGENET_DEFAULT_STD,
        # ),
        LazyCall(RandomErasing)(
            probability=0.25,
            mode="pixel",
            max_count=1,
            num_splits=0,
            device="cpu",
        ),
    ]
)
```
- 如果需要在别的服务器上跑ImageNet数据集，需要在`configs/vit_imagenet.py`下修改一下数据集的路径
```python
# Refine data path to imagenet
dataloader.train.dataset[0].root = "/path/to/imagenet"
dataloader.test[0].dataset.root = "/path/to/imagenet"

dataloader.train.dataset[0].root = "/dataset/extract"
dataloader.test[0].dataset.root = "/dataset/extract"
```

### CODE REVIEW

**Comment by BBuf** [comment]
# LiBai VIT性能问题

vs08上的8卡VIT测试结果。

| 框架                                               | Normalize怎么做             | 一个iter最高吞吐(imgs/s) |
| -------------------------------------------------- | --------------------------- | ------------------------ |
| OneFlow（经过cmakecache确认已经打开onednn） | ToTensor + Normalize        | 1230                     |
| OneFlow                                            | Normalize(numpy做)+ToTensor | 1738                     |

天和说的属实，LiBai的VIT DataLoader的Transform部分是性能瓶颈，上面的实验应该可以证明OneDNN优化之后的Normalize仍然比不过Numpy。

**Comment by rentainhe** [comment]
## LiBai ViT更多测试结果
经过晓雨路强的讨论，之前怀疑可能是因为/dataset这个盘的速度本身有点问题，所以这边将数据集替换为CIFAR100，然后将数据集放在/home目录下重新测试，其中数据增强对齐了ImageNet的数据增强，重新进行测试

| Model | Normalize Way | Throughput |
|:---:|:---:|:---:|
| ViT-Tiny with 1024 batchsize | Normalize在ToTensor之后 | 860 ~ 880 |
| ViT-Tiny with 1024 batchsize | Normalize在ToTensor之前 | 910 ~ 920 |
